### DISCUSSION TITLE
Rendering in MN versus VMD

### DISCUSSION THREAD

**Discussion by rbdavid** [BradyAJohnston/MolecularNodes#392] [General]
So, I'm trying to switch from using VMD to Blender/MN but I'm noticing some immediate issues.

- Slight differences in the NewCartoon representation used in VMD with the Cartoon Mesh in Blender. Maybe this shouldn't be surprising if Cartoon in Blender isn't trying to emulate NewCartoon. 
- There's some differences in residues' cartoon coloring (originating from #391) that I think stems from smoothing instead of individual colors for each residue's respective mesh section. 
- I've also struggled to recover the original camera orientation. 
- And the shadows are way harsher in Blender. 

Other than that, things are looking great!

From VMD:  
![model_quality](https://github.com/BradyAJohnston/MolecularNodes/assets/8484541/83df8428-7a71-4c6e-b951-b31187fd0c87)

From Blender: 
![blender_view](https://github.com/BradyAJohnston/MolecularNodes/assets/8484541/0f09a4ec-f405-4297-b890-6a165cb131a0)

Above, I've circled some of the differences in secondary structure visualizations in red. An example of the color smoothing issue is circled in green. 

So, my Molecular Nodes-relevant questions are: 
- What draw method is MN's Cartoon Mesh visualization using? Is Blender's Cartoon Mesh equivalent to either of VMD's NewCartoon or Cartoon drawing methods? 
- How do I turn off the smoothing in the cartoon mesh coloring?

I think the camera orientation issue just requires me to figure out how relative orientations are handled in Blender and how to control the camera position/viewport to mirror those I used in VMD. And I gotta figure out lighting types, positions, and strength to get a handle on the shadows. 

Any suggestions for getting a handle on all or some of these issues would be greatly appreciated!

**Comment by BradyAJohnston** [comment]
Thanks for the interest in using MN more. There are some things are are easily fixed and others that are less so.

>Slight differences in the NewCartoon representation used in VMD with the Cartoon Mesh in Blender. Maybe this shouldn't be surprising if Cartoon in Blender isn't trying to emulate NewCartoon.

I'm not trying to directly emulate NewCartoon from VMD, so I wouldn't expect them to be the same straight up. The differences in how they are styled is expected (how helices / beta sheets transition from one into the other etc). The Idea that I have is that there is a default, but that you can have access to the full spectrum of nobs to turn and tweak to get the style that you like. If you use `MN_utils_style_cartoon` node (`Molecular Nodes` -> `Utilities` -> `Cartoon Utils`) you can use the node that is used internally and exposes a lot more of the options to the user (I created a more simplified version that is easier to use initially).

That won't be able to get you all of the way to what you want though (at least currently). The smooth transition between secondary structure and ribbon isn't _currently_ possible, but in the next version of Blender (4.1 in a couple of months) this will be easier for me to achieve and so will be possible, and I'll include it as an option in the utils and higher level node.

If you have more suggestions about what you would like to be possible with the cartoon node, please do let me know and I can likely include in future releases. I sadly run up against limitations of Blender / Geometry Nodes in some cases (Cartoon was only possible a few months ago once I figured out some internal hacks) but these are constantly being improved which makes more styles possible.

>There's some differences in residues' cartoon coloring 

It's not something that I explicitly intended, but due to internal interpolation of colors inside of Geometry Nodes. I'll add an option to turn off interpolation in the next update, it's a simple fix and I am tracking it in #392

>I've also struggled to recover the original camera orientation.

If you can get the translation matrix from the camera, which you should be able to print in the VMD console, you should be able to then use that inside of Blender (as well as matching focal length etc).

> And the shadows are way harsher in Blender

This is all dependent on the lighting that you set up in your scene. The default scene isn't great with lighting unfortunately, but Blender has access to the full spectrum of potential lighting solutions. Because of this you can get some really nice lighting, but also because of this it can be difficult to _actually get_ really nice lighting. Including some more defaults with the add-on is on my to-do list, but it also requires me to figure them out myself as well.

**Reply by BradyAJohnston** [reply]
Additionally I forgot to mention, the differences where it is presence / absence of secondary structure, when importing currently the SS is detected with a very rudimentary algorithm from `biotite` which powers the parsing of `cif` files. This is a simplified version based mostly on the backbone, so it doesn't quite match up with the SS detection inside of VMD. 

This is something that I would love to address, but would also require myself to implement a SS detection algorithm. Something I would love to do but don't have the time to work on unfortunately. The models downloaded from the PDB use the SS encoded in the files, but those opened locally all have this detected version. I believe you can get the SS from the `cif` files, but `biotite` doesn't currently do that and again I haven't had the chance to implement it but it should be very possible.

**Reply by rbdavid** [reply]
Dope! Access to more nobs for the drawing methods is much appreciated. 

I'm certainly not wed to VMD's smoothing between secondary structure and unstructured loops. For MN, the abrupt end of helices and sheets removes ambiguity. Having arrows at the C-termini of the sheets does help visualize directionality; that may or may not be a difficult feature to add though. Now that I'm thinking about it, I'm not sure why alpha helices don't also have arrows at their C-terminus since there's a dipole moment across the helix that can be important for a protein's biophysics.  

In regards to the color interpolation fix, I'd be happy to implement that change. It might be a good first issue for a first-timer such as myself to close? 

As to the DSSP/STRIDE calculation, I can do some digging for python implementations of those. Alternatively, is it possible to add a switch that allows the user to point to and use a locally installed software (DSSP/STRIDE) to affect Blender-internal attributes? Like, all within Blender's python interface, the Blender-internal structural data is used as input to that locally-installed software, the output is captured and parsed, and the Blender visualization is created/updated.

**Reply by rbdavid** [reply]
The alternative was inspired by the biotite documentation that has this class that expects a locally installed version of DSSP: https://www.biotite-python.org/apidoc/biotite.application.dssp.DsspApp.html#biotite.application.dssp.DsspApp

**Reply by BradyAJohnston** [reply]
>Having arrows at the C-termini of the sheets does help visualize directionality; that may or may not be a difficult feature to add though.

This is already implemented, there should be a toggle option on the Style Cartoon node to enable / disable arrows (it should be enabled by default).  They aren't implemented fro AH, but I could add them as an option.

>In regards to the color interpolation fix, I'd be happy to implement that change. It might be a good first issue for a first-timer such as myself to close?

While I'd love to get you contributing, it's fairly deep inside of some of the geometry nodes node trees, so it would difficult to figure out what is going on and where best to implement. I just implemented it in #393 , and I'll create a new release with it shortly.

>Alternatively, is it possible to add a switch that allows the user to point to and use a locally installed software (DSSP/STRIDE) to affect Blender-internal attributes?

This is a good idea, and not one that I had considered yet. Certainly as biotite already supports potential interfacing with it, we could have an option in the preferences to select a `dssp` installation and use that for SS calculations.

**Reply by rbdavid** [reply]
> This is already implemented, there should be a toggle option on the Style Cartoon node to enable / disable arrows (it should be enabled by default). They aren't implemented fro AH, but I could add them as an option.

I'm still getting my sea-legs under me in regards to using Blender. I must have unselected the arrows checkbox early on in my testing and never looked back. Sorry for the extra pings about that. And thanks for your time! 

> While I'd love to get you contributing, it's fairly deep inside of some of the geometry nodes node trees, so it would difficult to figure out what is going on and where best to implement. I just implemented it in https://github.com/BradyAJohnston/MolecularNodes/issues/393 , and I'll create a new release with it shortly.

If I downloaded the .blend file you changed in #393 and replaced it in the molecularnodes addons directory tree, would that enable me to test out the non-interpolating coloring method? I want to improve the slides and videos for my job talk presentation by using blender  and I'm racing against a hidden clock at the moment 😛 This type of figure is key to one of my research projects. 

> This is a good idea, and not one that I had considered yet. Certainly as biotite already supports potential interfacing with it, we could have an option in the preferences to select a dssp installation and use that for SS calculations.

This kinda connects with my latest discussion thread ( #396 ) where I'm curious to know how to update/create attribute data associated with atoms. I'm gonna do some digging since I suspect this isn't a MN-specific question. I'll report back there with new insights.

**Reply by BradyAJohnston** [reply]
>If I downloaded the .blend file you changed in https://github.com/BradyAJohnston/MolecularNodes/issues/393 and replaced it in the molecularnodes addons directory tree, would that enable me to test out the non-interpolating coloring method?

This should work, yes. I have however just released 4.0.7 which now includes that as an update, along with `occupancy` and other improvements: https://github.com/BradyAJohnston/MolecularNodes/releases/tag/v4.0.7
